src/Lean/Widget/UserWidget.lean

```diff
@@ -0,0 +1,183 @@
+/-
+Copyright (c) 2022 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: E.W.Ayers
+-/
+import Lean.Widget.Basic
+import Lean.Data.Json
+import Lean.Environment
+import Lean.Server
+
+open Lean
+
+namespace Lean.Widget
+
+/-- A custom piece of code that is run on the editor client.
+
+The editor can use the `Lean.Widget.getWidgetSource` RPC method to
+get this object.
+
+See the [tutorial](doc/widgets.md) above this declaration for more information on
+how to use the widgets system.
+
+-/
+structure WidgetSource where
+  /-- Sourcetext of the code to run.-/
+  sourcetext : String
+  hash : String := toString <| hash sourcetext
+  deriving Inhabited, ToJson, FromJson
+
+namespace WidgetSource
+
+private def WidgetSourceRegistry := SimplePersistentEnvExtension
+    (String × WidgetSource)
+    (Std.RBMap String WidgetSource compare)
+
+instance : Inhabited (Std.RBMap String WidgetSource compare) := ⟨∅⟩
+instance : Inhabited (WidgetSourceRegistry) := inferInstanceAs (Inhabited (PersistentEnvExtension _ _ (List _ × _)))
+
+-- Mapping widgetSourceId to hash
+builtin_initialize widgetHashes : MapDeclarationExtension String ← mkMapDeclarationExtension `widgetHashes
+
+-- Mapping hash to wWidgetComponent
+builtin_initialize widgetSourceRegistry : WidgetSourceRegistry ←
```

Comment:
@leodemoura are there any performance implications of making an environment extension in this way?

Reply:
How many entries are you expecting this table to have?

Reply:
A low number, <20.